// ts-binary-serializer/dist/tests/largedata.variant.test.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var DataType_1 = require("../src/DataType");
var SerializeField_1 = require("../src/SerializeField");
var fs = __importStar(require("fs"));
var BinarySerializer_1 = require("../src/BinarySerializer");
var helper_test_1 = require("./helper.test");
describe("largedata-variant", function () {
    var buffer = fs.readFileSync("./testdata/prefab.json");
    var jstr = buffer.toString();
    var jobj = JSON.parse(jstr);
    it("case", function () {
        var sd = BinarySerializer_1.BinarySerializer.Serialize(jobj, GameObjectInfo);
        var _a = helper_test_1.estimateTime(function () { return BinarySerializer_1.BinarySerializer.Deserialize(sd, GameObjectInfo); }, 10), dt = _a[0], jobjd = _a[1];
        helper_test_1.verfiy(jobjd, jobj);
        console.log("variant:" + sd.byteLength, dt);
    });
});
var Vector2 = /** @class */ (function () {
    function Vector2(x, y) {
        if (x)
            this.x = x;
        else
            this.x = 0;
        if (y)
            this.y = y;
        else
            this.y = 0;
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Vector2.prototype, "x", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Vector2.prototype, "y", void 0);
    return Vector2;
}());
var Vector3 = /** @class */ (function () {
    function Vector3(x, y, z) {
        if (x)
            this.x = x;
        else
            this.x = 0;
        if (y)
            this.y = y;
        else
            this.y = 0;
        if (z)
            this.z = z;
        else
            this.z = 0;
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Vector3.prototype, "x", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Vector3.prototype, "y", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Vector3.prototype, "z", void 0);
    return Vector3;
}());
var Color = /** @class */ (function () {
    function Color() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Color.prototype, "r", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Color.prototype, "g", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Color.prototype, "b", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Color.prototype, "a", void 0);
    return Color;
}());
var Frame = /** @class */ (function () {
    function Frame() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Frame.prototype, "time", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Frame.prototype, "val", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], Frame.prototype, "str", void 0);
    return Frame;
}());
var Sprite = /** @class */ (function () {
    function Sprite() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], Sprite.prototype, "spriteName", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, false, Color)
    ], Sprite.prototype, "spriteTint", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, false, Vector2)
    ], Sprite.prototype, "spritePivot", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Bool)
    ], Sprite.prototype, "flipX", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Bool)
    ], Sprite.prototype, "flipY", void 0);
    return Sprite;
}());
var State = /** @class */ (function () {
    function State(info) {
        this.name = info.name;
        this.aniName = info.aniName;
        this.nextStates = info.nextStates;
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], State.prototype, "name", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], State.prototype, "aniName", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String, true)
    ], State.prototype, "nextStates", void 0);
    return State;
}());
var Curve = /** @class */ (function () {
    function Curve(info) {
        this.framesDic = {};
        // this.aim = info.aim;
        // this.type = info.type;
        // this.frames = info.frames;
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], Curve.prototype, "aim", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], Curve.prototype, "type", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, Frame)
    ], Curve.prototype, "frames", void 0);
    return Curve;
}());
var Event = /** @class */ (function () {
    function Event() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Event.prototype, "time", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], Event.prototype, "functionName", void 0);
    return Event;
}());
var Clip = /** @class */ (function () {
    function Clip(info) {
        this.name = info.name;
        this.curves = new Array();
        for (var i = 0; i < info.curves.length; i++)
            this.curves.push(new Curve(info.curves[i]));
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], Clip.prototype, "name", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, Curve)
    ], Clip.prototype, "curves", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, Event)
    ], Clip.prototype, "events", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Bool)
    ], Clip.prototype, "loop", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Varint32)
    ], Clip.prototype, "frameCount", void 0);
    return Clip;
}());
var Animation = /** @class */ (function () {
    function Animation() {
        this.defaultState = "";
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, Clip)
    ], Animation.prototype, "animClips", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, State)
    ], Animation.prototype, "states", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], Animation.prototype, "defaultState", void 0);
    return Animation;
}());
var Text = /** @class */ (function () {
    function Text() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], Text.prototype, "text", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, false, Color)
    ], Text.prototype, "txtTint", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Varint32)
    ], Text.prototype, "size", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Varint32)
    ], Text.prototype, "pivot", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Text.prototype, "width", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Text.prototype, "height", void 0);
    return Text;
}());
var Node = /** @class */ (function () {
    function Node() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], Node.prototype, "name", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, false, Vector3)
    ], Node.prototype, "position", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, false, Vector3)
    ], Node.prototype, "scale", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], Node.prototype, "rotation", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Bool)
    ], Node.prototype, "active", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String, true)
    ], Node.prototype, "children", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, false, Sprite)
    ], Node.prototype, "sprite", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, false, Text)
    ], Node.prototype, "txt", void 0);
    return Node;
}());
var GameObjectInfo = /** @class */ (function () {
    function GameObjectInfo() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], GameObjectInfo.prototype, "root", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, Node)
    ], GameObjectInfo.prototype, "nodes", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, false, Animation)
    ], GameObjectInfo.prototype, "anim", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String, true)
    ], GameObjectInfo.prototype, "components", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, Vector2)
    ], GameObjectInfo.prototype, "colliderPoly", void 0);
    return GameObjectInfo;
}());
var SprExInfo = /** @class */ (function () {
    function SprExInfo() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], SprExInfo.prototype, "name", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, false, Vector2)
    ], SprExInfo.prototype, "pivot", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], SprExInfo.prototype, "pixelPerUnit", void 0);
    return SprExInfo;
}());
var SprExtraInfo = /** @class */ (function () {
    function SprExtraInfo() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, SprExInfo)
    ], SprExtraInfo.prototype, "infos", void 0);
    return SprExtraInfo;
}());
var StaticItemVO = /** @class */ (function () {
    function StaticItemVO() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], StaticItemVO.prototype, "m_entity", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], StaticItemVO.prototype, "m_type", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], StaticItemVO.prototype, "m_roomID", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], StaticItemVO.prototype, "m_status", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], StaticItemVO.prototype, "m_dir", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], StaticItemVO.prototype, "m_x", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], StaticItemVO.prototype, "m_y", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], StaticItemVO.prototype, "m_width", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], StaticItemVO.prototype, "m_height", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], StaticItemVO.prototype, "m_zOrder", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Bool)
    ], StaticItemVO.prototype, "m_block", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Bool)
    ], StaticItemVO.prototype, "m_enable", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Bool)
    ], StaticItemVO.prototype, "m_visible", void 0);
    return StaticItemVO;
}());
var TileWallVO = /** @class */ (function () {
    function TileWallVO() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileWallVO.prototype, "m_type", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileWallVO.prototype, "m_entity", void 0);
    return TileWallVO;
}());
var TileVO = /** @class */ (function () {
    function TileVO() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileVO.prototype, "m_x", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileVO.prototype, "m_y", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileVO.prototype, "m_btEntity", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileVO.prototype, "m_upEntity", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Bool)
    ], TileVO.prototype, "m_block", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileVO.prototype, "m_blockID", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileVO.prototype, "m_regionID", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, TileWallVO)
    ], TileVO.prototype, "m_walls", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileVO.prototype, "m_photoRegionID", void 0);
    return TileVO;
}());
var MapEntityVO = /** @class */ (function () {
    function MapEntityVO() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], MapEntityVO.prototype, "m_prefabName", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.String)
    ], MapEntityVO.prototype, "m_imageName", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], MapEntityVO.prototype, "m_offsetX", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], MapEntityVO.prototype, "m_offsetY", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], MapEntityVO.prototype, "m_pivotX", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], MapEntityVO.prototype, "m_pivotY", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], MapEntityVO.prototype, "m_scale", void 0);
    return MapEntityVO;
}());
var TileMapVO = /** @class */ (function () {
    function TileMapVO() {
    }
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileMapVO.prototype, "m_width", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileMapVO.prototype, "m_height", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileMapVO.prototype, "m_tileWidth", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32)
    ], TileMapVO.prototype, "m_tileHeight", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, TileVO)
    ], TileMapVO.prototype, "m_tiles", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, StaticItemVO)
    ], TileMapVO.prototype, "m_staticItems", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Object, true, MapEntityVO)
    ], TileMapVO.prototype, "m_mapEntities", void 0);
    __decorate([
        SerializeField_1.SerializeField(DataType_1.DataType.Float32, true)
    ], TileMapVO.prototype, "m_tilesID", void 0);
    return TileMapVO;
}());
