// ts-binary-serializer/dist/src/DataType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DataType;
(function (DataType) {
    DataType[DataType["Null"] = 0] = "Null";
    DataType[DataType["Float32"] = 1] = "Float32";
    DataType[DataType["Float64"] = 2] = "Float64";
    DataType[DataType["Int32"] = 3] = "Int32";
    DataType[DataType["Int16"] = 4] = "Int16";
    DataType[DataType["Int8"] = 5] = "Int8";
    DataType[DataType["Uint32"] = 6] = "Uint32";
    DataType[DataType["Uint16"] = 7] = "Uint16";
    DataType[DataType["Uint8"] = 8] = "Uint8";
    DataType[DataType["String"] = 9] = "String";
    DataType[DataType["Bool"] = 10] = "Bool";
    DataType[DataType["Object"] = 11] = "Object";
    DataType[DataType["Float16"] = 12] = "Float16";
    DataType[DataType["Map"] = 13] = "Map";
    DataType[DataType["Varint32"] = 14] = "Varint32";
    DataType[DataType["UVarint32"] = 15] = "UVarint32";
    DataType[DataType["TypedArray"] = 16] = "TypedArray";
})(DataType = exports.DataType || (exports.DataType = {}));
