// ts-binary-serializer/dist/src/TypeMetaClass.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TypeMetaProperty = /** @class */ (function () {
    function TypeMetaProperty(key, datatype, isary, pclass) {
        if (isary === void 0) { isary = false; }
        this.isArray = false;
        this.key = key;
        this.datatype = datatype;
        this.isArray = isary;
        this.pclass = pclass;
    }
    return TypeMetaProperty;
}());
exports.TypeMetaProperty = TypeMetaProperty;
function sort(a, b) {
    return a.key.localeCompare(b.key);
}
var TypeMetaClass = /** @class */ (function () {
    function TypeMetaClass() {
        this.m_needSort = true;
    }
    Object.defineProperty(TypeMetaClass.prototype, "prototype", {
        get: function () {
            return this.m_prototype;
        },
        set: function (v) {
            this.m_prototype = v;
            this.pname = this.protoName;
        },
        enumerable: true,
        configurable: true
    });
    TypeMetaClass.prototype.addProperty = function (k, t, isary) {
        if (isary === void 0) { isary = false; }
        var p = this.properties;
        for (var i = 0, len = p.length; i < len; i++) {
            if (p[i].key == k)
                return;
        }
        p.push(new TypeMetaProperty(k, t, isary));
        this.m_needSort = true;
    };
    Object.defineProperty(TypeMetaClass.prototype, "protoName", {
        get: function () {
            return this.prototype.constructor.name;
        },
        enumerable: true,
        configurable: true
    });
    TypeMetaClass.prototype.sortProperty = function () {
        if (!this.m_needSort)
            return;
        this.properties.sort(function (a, b) { return sort(a, b); });
        this.m_needSort = false;
        return this;
    };
    return TypeMetaClass;
}());
exports.TypeMetaClass = TypeMetaClass;
